db/db.go

```diff
@@ -76,20 +78,70 @@ func Init(config config.DBConfig, debug bool) *Database {
 	if debug {
 		db.Config.Logger.LogMode(logger.Info)
 	}
-	return &Database{db}
+
+	d := &Database{db}
+	if err = d.MigrateLineage(); err != nil {
+		log.Fatalf("Lineage migration failed: %v\n", err)
+	}
+
+	return d
+}
+
+// MigrateLineage is a migration function to update db and its data to the
+// new lineage db scheme. It will update State table data, delete "lineage" column
+// and add corresponding Lineage entries
+func (db *Database) MigrateLineage() error {
+	if db.Migrator().HasColumn(&types.State{}, "lineage") {
+		states := db.ListStates()
+		for _, stPath := range states {
+			// Recover State from db for update
+			var st types.State
+			res := db.First(&st, types.State{Path: stPath})
+			if errors.Is(res.Error, gorm.ErrRecordNotFound) {
+				return fmt.Errorf("State not found in db")
+			}
+
+			if err := db.UpdateState(st); err != nil {
+				return fmt.Errorf("Failed to update %s state during lineage migration: %v", stPath, err)
+			}
+		}
+
+		// Custom migration rules
+		if err := db.Migrator().DropColumn(&types.State{}, "lineage"); err != nil {
+			return fmt.Errorf("Failed to drop lineage column during migration: %v", err)
+		}
+	}
+
+	return nil
 }
 
 type attributeValues map[string]interface{}
 
 func (db *Database) stateS3toDB(sf *statefile.File, path string, versionID string) (st types.State) {
 	var version types.Version
 	db.First(&version, types.Version{VersionID: versionID})
-	st = types.State{
-		Path:      path,
-		Version:   version,
-		TFVersion: sf.TerraformVersion.String(),
-		Serial:    int64(sf.Serial),
-		Lineage:   sf.Lineage,
+
+	// Check if the associated lineage is already present in lineages table
+	// If so, it recovers its ID otherwise it inserts it at the same time as the state
+	var lineage types.Lineage
+	if errors.Is(db.First(&lineage, types.Lineage{Value: sf.Lineage}).Error, gorm.ErrRecordNotFound) {
+		st = types.State{
+			Path:      path,
+			Version:   version,
+			TFVersion: sf.TerraformVersion.String(),
+			Serial:    int64(sf.Serial),
+			Lineage: types.Lineage{
+				Value: sf.Lineage,
+			},
+		}
+	} else {
+		st = types.State{
+			Path:      path,
+			Version:   version,
+			TFVersion: sf.TerraformVersion.String(),
+			Serial:    int64(sf.Serial),
+			LineageID: lineage.ID,
+		}
```

Comment:
What happens if there's an error but it's not `gorm.ErrRecordNotFound`?

Reply:
On the Gorm.v2 documentation, they only evoke this error for db.First, so I started from the principle that there could be no other ... An interpretation that may be wrong, however, what do you think?
By the way, putting a more generic error handling here will not cost anything and will remove the doubt :monocle_face:

Reply:
I think it'd be safer to have 3 cases:

* `err == nil`
* `err` is `gorm.ErrRecordNotFound`
* else raise an `Unknown error`